commit 15c052b9fe72c33518615a69ccdd7a71779730a5 (refs/changes/68/4468/3)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2018-07-23T22:53:51+00:00 (1 year, 3 months ago)
    
    MORAY-483 Prevent new queries when Postgres client has emitted an error

diff --git a/lib/pg.js b/lib/pg.js
index 02b25cc..4a16e74 100644
--- a/lib/pg.js
+++ b/lib/pg.js
@@ -247,6 +247,18 @@ PGClient.prototype.query = function clientQuery(sql, args) {
     var self = this;
     var timer;
 
+    if (self._moray_had_err !== null) {
+        /*
+         * If we've hit a connection error, then we need to make sure to
+         * avoid calling the node-postgres .query() method, since it won't
+         * emit anything.
+         */
+        setImmediate(function () {
+            res.emit('error', self._moray_had_err);
+        });
+        return res;
+    }
+
     /*
      * Moray will periodically issue the query 'SELECT NOW() AS when' to
      * available PGClients as a health checking facility (see pgCheck).
@@ -274,7 +286,7 @@ PGClient.prototype.query = function clientQuery(sql, args) {
             return ([reqid, sql, err.toString()]);
         });
 
-        done('error', err);
+        setImmediate(done, 'error', err);
     }
 
     req = new pg.Query(sql, args);
@@ -377,7 +389,7 @@ PGClient.prototype.query = function clientQuery(sql, args) {
         args: args
     }, 'pg.query: started');
 
-    return (res);
+    return res;
 };
 
 
